From 924e11f95f33c23a2e68c2d38892b6aed617882f Mon Sep 17 00:00:00 2001
From: zhengq <qi.zheng@intel.com>
Date: Wed, 8 Oct 2014 22:18:22 +0800
Subject: [PATCH 1/1] PCI/PM: Enable runtime PM for CHT PCIe ports.

From commit de7d5f729c72638f41d7c17487bccb1c570ff144, there are
some problems when enabling runtime PM for PCIe ports on some
chipsets.
Try enable this runtime PM support on CHT only while keep it
disabled on all other platforms for GMIN CHT development.

Change-Id: I5f9ee0732eeb4daba91fdc2cae007a888a2cbad5
Signed-off-by: zhengq <qi.zheng@intel.com>
---
 drivers/pci/pcie/portdrv_pci.c |   20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/drivers/pci/pcie/portdrv_pci.c b/drivers/pci/pcie/portdrv_pci.c
index 0d8fdc48e642..7fb614eda35d 100644
--- a/drivers/pci/pcie/portdrv_pci.c
+++ b/drivers/pci/pcie/portdrv_pci.c
@@ -185,6 +185,17 @@ static const struct dev_pm_ops pcie_portdrv_pm_ops = {
 #endif /* !PM */
 
 /*
+ * PCIe port runtime suspend is broken for some chipsets, so use a
+ * enable list to enable runtime PM only for good chipsets.
+ */
+static const struct pci_device_id port_runtime_pm_enable_list[] = {
+	/*Cherryview*/
+	{PCI_DEVICE(PCI_VENDOR_ID_INTEL, 0x22c8),},
+
+	{ /* end: all zeroes */ }
+};
+
+/*
  * pcie_portdrv_probe - Probe PCI-Express port devices
  * @dev: PCI-Express port device being probed
  *
@@ -217,11 +228,20 @@ static int pcie_portdrv_probe(struct pci_dev *dev,
 	 * it by default.
 	 */
 	dev->d3cold_allowed = false;
+	if (pci_match_id(port_runtime_pm_enable_list, dev)) {
+		pm_runtime_put_noidle(&dev->dev);
+		pm_runtime_allow(&dev->dev);
+	}
+
 	return 0;
 }
 
 static void pcie_portdrv_remove(struct pci_dev *dev)
 {
+	if (pci_match_id(port_runtime_pm_enable_list, dev)) {
+		pm_runtime_forbid(&dev->dev);
+		pm_runtime_get_noresume(&dev->dev);
+	}
 	pcie_port_device_remove(dev);
 }
 
-- 
1.7.9.5

